components/node/node/api.go

```diff
@@ -91,73 +90,49 @@ func (n *nodeAPI) OutputAtBlock(ctx context.Context, number hexutil.Uint64) (*et
 	recordDur := n.m.RecordRPCServerRequest("kroma_outputAtBlock")
 	defer recordDur()
 
-	ref, nextRef, status, err := n.dr.BlockRefsWithStatus(ctx, uint64(number))
+	output, err := n.fetchOutputAtBlock(ctx, number)
 	if err != nil {
-		return nil, fmt.Errorf("failed to get L2 block ref with sync status: %w", err)
+		return nil, err
 	}
 
-	head, err := n.client.InfoByHash(ctx, ref.Hash)
+	return output, nil
+}
+
+func (n *nodeAPI) OutputWithProofAtBlock(ctx context.Context, number hexutil.Uint64) (*eth.OutputResponse, error) {
+	recordDur := n.m.RecordRPCServerRequest("kroma_outputWithProofAtBlock")
+	defer recordDur()
+
+	output, err := n.fetchOutputAtBlock(ctx, number)
 	if err != nil {
-		return nil, fmt.Errorf("failed to get L2 block by hash %s: %w", ref, err)
-	}
-	if head == nil {
-		return nil, ethereum.NotFound
+		return nil, err
 	}
 
-	proof, err := n.client.GetProof(ctx, predeploys.L2ToL1MessagePasserAddr, []common.Hash{}, ref.Hash.String())
+	nextHead, nextTxs, err := n.client.InfoAndTxsByHash(ctx, output.NextBlockRef.Hash)
 	if err != nil {
-		return nil, fmt.Errorf("failed to get contract proof at block %s: %w", ref, err)
-	}
-	if proof == nil {
-		return nil, fmt.Errorf("proof %w", ethereum.NotFound)
-	}
-	// make sure that the proof (including storage hash) that we retrieved is correct by verifying it against the state-root
-	if err := proof.Verify(head.Root()); err != nil {
-		n.log.Error("invalid withdrawal root detected in block", "stateRoot", head.Root(), "blocknum", number, "msg", err)
-		return nil, fmt.Errorf("invalid withdrawal root hash, state root was %s: %w", head.Root(), err)
+		return nil, fmt.Errorf("failed to get L2 block by hash %s: %w", output.NextBlockRef, err)
 	}
+	nextBlock := nextHead.Header()
 
-	var l2OutputRootVersion eth.Bytes32
-	var l2OutputRoot eth.Bytes32
-	if n.config.IsBlue(head.Time()) {
-		l2OutputRootVersion = rollup.V1
-		l2OutputRoot, err = rollup.ComputeL2OutputRootV1(&bindings.TypesOutputRootProof{
-			Version:                  l2OutputRootVersion,
-			StateRoot:                head.Root(),
-			MessagePasserStorageRoot: proof.StorageHash,
-			BlockHash:                head.Hash(),
-			NextBlockHash:            nextRef.Hash,
-		})
-	} else {
-		l2OutputRootVersion = rollup.V0
-		l2OutputRoot, err = rollup.ComputeL2OutputRootV0(&bindings.TypesOutputRootProof{
-			Version:                  l2OutputRootVersion,
-			StateRoot:                head.Root(),
-			MessagePasserStorageRoot: proof.StorageHash,
-			BlockHash:                head.Hash(),
-		})
+	accountResult, err := n.client.GetProof(ctx, predeploys.L2ToL1MessagePasserAddr, []common.Hash{}, output.BlockRef.Hash.String())
```

Comment:
`n.fetchOutputAtBlock` also call `GetProof` of `predeploys.L2ToL1MessagePasserAddr`, so can we remove this repeated call?

Reply:
I marked it as TODO, and will apply it in future :)